### ISSUE TITLE
useRouterMatch returns new object on each render

### ISSUE THREAD

**Issue by bkazi** [remix-run/react-router#7059]
## Version

5.1.2

## Test Case

https://codesandbox.io/s/eloquent-wave-vlixh

## Steps to reproduce

CodeSanbox has a reproducible case

## Expected Behavior

`useRouteMatch` hook returns a match object, when combined with a `useEffect` it is expected that the effect will run only when the window location or path parameters change

## Actual Behavior

`useRouteMatch` hook returns a match object, when combined with a `useEffect` with the match as dependency the effect is triggered on every state change.
The codesandbox example performs a state update on a timeout and logs `match` whenever it is changed. The logs show that on every update match has changed even if the path has not. Memoizing the `match` object on `window.location.pathname` prevents this and has the expected behaviour

[This line](https://github.com/ReactTraining/react-router/blob/e81dfa2d01937969ee3f9b1f33c9ddd319f9e091/packages/react-router/modules/hooks.js#L54) in the hook returns the result of [`matchPath`](https://github.com/ReactTraining/react-router/blob/e81dfa2d01937969ee3f9b1f33c9ddd319f9e091/packages/react-router/modules/matchPath.js#L55) which creates a new object every time it is executed.
Possible solution would be to do something like
```js
return path ? 
  useMemo(() => matchPath(location.pathname, path), [location.pathname, path]) :
  match
```

**Comment by timdorr** [comment]
I think that makes sense. Can you submit a PR?

**Comment by pavelkuznetsov** [comment]
Hi! There was kind of similar question about new `match` object created on each render - https://github.com/ReactTraining/react-router/issues/5099 . 

The issue was closed because it is intentional, but now, with the `useRouteMatch`, the answer is that memoization could help.

@timdorr, is there any difference in these cases? Perhaps, `Route` could pass memoized `match` to its `component` too? It seems like it could help with kind of unnecessary re-rerenders described in the issue I linked above. 

P.S. It could really help people like me, which migrating from RR3 to the last one and still use redux and react-redux. After migration we found out that our 'connected pure components' from react-redux `connect` HOC, which are passed to `Route`'s `component` prop, aren't pure anymore.

**Comment by timdorr** [comment]
Yes, that was when using lifecycle methods and a class component. You could "memoize" by way of shouldComponentUpdate to prevent a render. With Hooks, you're already in a render cycle, so it's not possible to bail out. That's why useMemo/useCallback even exist, so you can safely memoize values directly.

**Comment by pavelkuznetsov** [comment]
Thanks for quick reply!

As i understood, you don't think that it makes sense if `Route` could memoize `match` object out of the box (like it was described for `useRouteMatch`) in order to recalculate and pass new `match` object to component props only when the location or path parameters have changed?

**Comment by LoiKos** [comment]
I face the same problem with `useRouteMatch`  today what is the status of this issue ?

**Comment by lauterry** [comment]
same issue with useLocation

```
const location = useLocation();

...

useEffect(() => {
	console.log(location)
}, [location]);
```

Go infinite loop

**Comment by LoiKos** [comment]
I tried to reproduce infinite loop in my code for `useLocation` and i wasn't able to reproduce.

**Comment by lauterry** [comment]
@LoiKos Sorry, I have just noticed that my code was not complete. I was calling history.replace to reset the location state

**Comment by LoiKos** [comment]
Yes i can reproduce the infinite loop which is normal btw. You try to add a state with ` history.replace({ pathname: "/currentPath", state: { comeFromError: undefined } });` but as state is a part of location you create an infinite loop. If you only want to create the state the first time you came on `/currentPath` then you have to listen to `location.pathname` as you find. 

In my opinion the best practice with hooks dependencies is to be as accurate as possible on which variable you want to listen to. That avoid unnecessary operation and bugs that come from an object getting updated but not the part you want to look after. 

I think you also forget to add `history` in your dependencies

**Comment by bluefire2121** [comment]
still looking for an update. Can this be memoized inside the package?

Scenario:
When using useRouteMatch and a primative property of match.params, it can't be passed into the dependency array because that property doesn't exist if there isn't a match.

**Comment by olejorgenb** [comment]
A bit off topic, but simply using `useLocation` inside a memo'ed component will cause a rerendering if any Route above the component rerenders: https://codesandbox.io/s/festive-greider-8did7

The location value isn't changed so it is trival to work-around, but quite a gothca IMO. https://github.com/facebook/react/issues/15156#issuecomment-474590693 for a related discussion.

**Comment by richardscarrott** [comment]
If you want to perform side-effects on location change, then I think you can prob just rely on `location`:

```js
const location = useLocation();
const match = useMatch();
useEffect(() => {
  // The location has changed
  nonIdempotentSideEffect(match.params.foo);
}, [location])
```

Or I think you could make match stable in user land like this:
```js
const useStableRouteMatch = () => {
  const _match = useRouteMatch();
  const location = useLocation();
  const [match, setMatch] = useState(_match);
  useEffect(() => {
    setMatch(_match);
  }, [location]);
  return match;
};

// ...

const match = useStableRouteMatch();
useEffect(() => {
  // The location has changed
  nonIdempotentSideEffect(match.params.foo);
}, [match]);
```

However, I don't think using `React.useMemo` would be safe unless you were merely trying to improve render performance as there's no guarantee React won't just throw away the memo'd value, e.g. **don't** do this:

```js
const useMemoRouteMatch = () => {
  const match = useRouteMatch();
  const location = useLocation();
  return useMemo(() => match, [location]);
}

// ...

const match = useMemoRouteMatch();
useEffect(() => {
  // The location has changed *or react has discarded the memo'd match value*
  nonIdempotentSideEffect(match.params.foo);
}, [match]);
```

**Comment by vladshcherbin** [comment]
Same for other hooks like `useParams`. It triggers a rerender on every route change, even if there are no params in url. So basically using hooks = trigger app rerender on every route change.

Maybe passing props you want to subscribe to into this hooks may help.

**Comment by andreybs11** [comment]
Nested routes trigger the infinite re-render loop. Would be great to memoize the hooks +1

**Comment by filipealonso** [comment]
I also get an infinite loop with useHistory and useLocation.

Steps:
1. Open URL with a query string and load a specific a route
2. Use useLocation to detect query string
3. If there is query string use useHistory to redirect to same URL without the query string
4. The URL changes in the browser but useLocation keeps detecting the previous query string and triggering step 3 which keeps redirecting to URL without query string (step 4) and so on...

**Comment by dtokarczyk** [comment]
+1

**Comment by Themonkey180** [comment]
I am still having this issue, is there a update on it?

**Comment by dwelle** [comment]
Unclear on what's the stance on `<Router match>` [object memoization](https://github.com/ReactTraining/react-router/issues/7059#issuecomment-564141131) --- should we create a new issue for that?

(Haven't checked if v6 changed the behavior already.)

**Comment by Powersource** [comment]
In the meantime this `history.listen` function works pretty well for me https://github.com/ReactTraining/history/blob/v4/docs/GettingStarted.md#listening

**Comment by Powersource** [comment]
An example of how I use `history.listen`

```js
function useListenHistory(fn) {
  const [state, setState] = useState(fn());
  const history = useHistory();

  useEffect(() => history.listen(() => {
    const newState = fn();

    if (newState !== state) {
      setState(newState);
    }
  }));

  return state;
}

export function useAtHome() {
  const getAtHome = () => window.location.pathname === '/';
  return useListenHistory(getAtHome);
}
```

**Comment by mafergus** [comment]
This is causing all kind of unnecessary re-rendering, we've had to write our own set of wrapper functions to compensate

**Comment by jahglow** [comment]
> An example of how I use `history.listen`
> 
> ```js
> function useListenHistory(fn) {
>   const [state, setState] = useState(fn());
>   const history = useHistory();
> 
>   useEffect(() => history.listen(() => {
>     const newState = fn();
> 
>     if (newState !== state) {
>       setState(newState);
>     }
>   }));
> 
>   return state;
> }
> 
> export function useAtHome() {
>   const getAtHome = () => window.location.pathname === '/';
>   return useListenHistory(getAtHome);
> }
> ```

your example has zero memoization itself neither in `getAtHome` nor in `useEffect` and guess what, your useEffect is triggered on every render. Add deps to useEffect and wrap `getAtHome` with useCallback.

**Comment by jonkoops** [comment]
I've opened up a PR to start a discussion on how this issue could be resolved in https://github.com/ReactTraining/react-router/pull/7822.

**Comment by zac-robinson** [comment]
Is there any update on the PR?

**Comment by chaance** [comment]
The update is that I'll be reviewing in the next week or two. This one is on the radar.

**Comment by jonkoops** [comment]
Created a new PR (#8431) to resolve this issue.

**Comment by timdorr** [comment]
That's now released in 6.1.0! :)
